TEMPORARY bring-up probes. Drop this patch once the port boots: it only prints.

--- a/bsd/kern/sys_generic.c
+++ b/bsd/kern/sys_generic.c
@@ -701,6 +701,11 @@
 int
 write_nocancel(struct proc *p, struct write_nocancel_args *uap, user_ssize_t *retval)
 {
+#if defined(KEBAB)
+	/* Bring-up probe: is userland reaching the console at all? */
+	printf("KEBAB: write fd=%d n=%llu pid=%d\n", uap->fd,
+	    (unsigned long long)uap->nbyte, proc_pid(p));
+#endif
 	return write_internal(p, uap->fd, uap->cbuf, uap->nbyte, (off_t)-1, 0,
 	           NULL, retval);
 }
--- a/bsd/kern/kern_descrip.c
+++ b/bsd/kern/kern_descrip.c
@@ -2718,6 +2718,15 @@
 	    fopen.o_pathname, has_entitlement ? &context : vfs_context_current());
 	nd->ni_dvp = vp;
 
+#if defined(KEBAB)
+	{
+		/* Bring-up probe: the first thing PID 1 does is open its console. */
+		char kpath[64] = "?";
+		(void)copyinstr(fopen.o_pathname, kpath, sizeof(kpath), NULL);
+		printf("KEBAB: open \"%s\" flags=0x%x pid=%d\n",
+		    kpath, fopen.o_flags, proc_pid(p));
+	}
+#endif
 	error = open1(has_entitlement ? &context : vfs_context_current(),
 	    nd, fopen.o_flags, va, NULL, NULL, retval, AUTH_OPEN_NOAUTHFD);
 
--- a/bsd/kern/kern_exec.c
+++ b/bsd/kern/kern_exec.c
@@ -962,6 +962,10 @@
 get_teamid_for_shared_region(struct image_params *imgp)
 {
 	assert(imgp->ip_vp != NULL);
+#if defined(KEBAB)
+	printf("KEBAB: namei ok, vnode %p size %lld\n",
+	    (void *)imgp->ip_vp, (long long)imgp->ip_vp->v_size);
+#endif
 
 	const char *ret = csvnode_get_teamid(imgp->ip_vp, imgp->ip_arch_offset);
 	if (ret) {
@@ -1160,6 +1164,9 @@
 static int
 exec_mach_imgact(struct image_params *imgp)
 {
+#if defined(KEBAB)
+	printf("KEBAB: exec_mach_imgact enter\n");
+#endif
 	struct mach_header *mach_header = (struct mach_header *)imgp->ip_vdata;
 	proc_t                  p = vfs_context_proc(imgp->ip_vfs_context);
 	int                     error = 0;
@@ -1314,7 +1321,13 @@
 	/*
 	 * Actually load the image file we previously decided to load.
 	 */
+#if defined(KEBAB)
+	printf("KEBAB: calling load_machfile\n");
+#endif
 	lret = load_machfile(imgp, mach_header, thread, &map, &load_result);
+#if defined(KEBAB)
+	printf("KEBAB: load_machfile -> %d\n", (int)lret);
+#endif
 	if (lret != LOAD_SUCCESS) {
 		error = load_return_to_errno(lret);
 
@@ -1364,6 +1377,9 @@
 
 	proc_lock(p);
 	p->p_cputype = imgp->ip_origcputype;
+#if defined(KEBAB)
+	printf("KEBAB: exec P1 post-load\n");
+#endif
 	p->p_cpusubtype = imgp->ip_origcpusubtype;
 	proc_setplatformdata(p, load_result.ip_platform, load_result.lr_min_sdk, load_result.lr_sdk);
 	exec_setup_tpro(imgp, &load_result);
@@ -1378,6 +1394,9 @@
 		assert(!vm_map_is_alien(map));
 	}
 #endif /* XNU_TARGET_OS_OSX */
+#if defined(KEBAB)
+	printf("KEBAB: exec P2 proc unlock\n");
+#endif
 	proc_unlock(p);
 
 	/*
@@ -1608,6 +1627,10 @@
 	 * vm_map_switch.
 	 */
 	old_map = swap_task_map(task, thread, map);
+#if defined(KEBAB)
+	printf("KEBAB: exec P3 swap map\n");
+#endif
+
 #if MACH_ASSERT
 	/*
 	 * Reset the pmap's process info to prevent ledger checks
@@ -1618,8 +1641,15 @@
 #endif /* MACH_ASSERT */
 	vm_map_deallocate(old_map);
 	old_map = NULL;
+#if defined(KEBAB)
+	printf("KEBAB: exec P4 old map freed\n");
+#endif
+
 
 	lret = activate_exec_state(task, p, thread, &load_result);
+#if defined(KEBAB)
+	printf("KEBAB: exec P5 exec state active\n");
+#endif
 	if (lret != KERN_SUCCESS) {
 		KERNEL_DEBUG_CONSTANT(BSDDBG_CODE(DBG_BSD_PROC, BSD_PROC_EXITREASON_CREATE) | DBG_FUNC_NONE,
 		    proc_getpid(p), OS_REASON_EXEC, EXEC_EXIT_REASON_ACTV_THREADSTATE, 0, 0);
@@ -1662,6 +1692,9 @@
 
 		goto badtoolate;
 	}
+#if defined(KEBAB)
+	printf("KEBAB: exec P6 stack created\n");
+#endif
 
 	/*
 	 * The load result will have already been munged by AMFI to include the
@@ -1703,6 +1736,9 @@
 	    load_result.lr_sdk);
 
 	error = exec_add_apple_strings(imgp, &load_result); /* copies out main thread port */
+#if defined(KEBAB)
+	printf("KEBAB: exec P7 apple strings -> %d\n", error);
+#endif
 
 	if (error) {
 		KERNEL_DEBUG_CONSTANT(BSDDBG_CODE(DBG_BSD_PROC, BSD_PROC_EXITREASON_CREATE) | DBG_FUNC_NONE,
@@ -1718,6 +1754,9 @@
 
 	/* Switch to target task's map to copy out strings */
 	old_map = vm_map_switch(get_task_map(task));
+#if defined(KEBAB)
+	printf("KEBAB: exec P8 map switched\n");
+#endif
 
 	if (load_result.unixproc) {
 		user_addr_t     ap;
@@ -1743,6 +1782,9 @@
 		}
 		/* Set the stack */
 		thread_setuserstack(thread, ap);
+#if defined(KEBAB)
+		printf("KEBAB: exec P9 stack set\n");
+#endif
 	}
 
 	if (load_result.dynlinker || load_result.is_rosetta) {
@@ -6879,10 +6921,20 @@
 	mach_vm_offset_t scratch_addr = 0;
 	mach_vm_size_t map_page_size = vm_map_page_size(map);
 
+#if defined(KEBAB)
+	printf("KEBAB: load_init_program enter (map %p, page 0x%llx)\n",
+	    (void *)map, (unsigned long long)map_page_size);
+#endif
 	(void) mach_vm_allocate_kernel(map, &scratch_addr, map_page_size, VM_FLAGS_ANYWHERE, VM_KERN_MEMORY_NONE);
+#if defined(KEBAB)
+	printf("KEBAB: scratch allocated at 0x%llx\n", (unsigned long long)scratch_addr);
+#endif
 #if CONFIG_MEMORYSTATUS
 	(void) memorystatus_init_at_boot_snapshot();
 #endif /* CONFIG_MEMORYSTATUS */
+#if defined(KEBAB)
+	printf("KEBAB: boot snapshot recorded\n");
+#endif
 
 #if DEBUG || DEVELOPMENT
 	/* Check for boot-arg suffix first */
--- a/bsd/dev/arm/systemcalls.c
+++ b/bsd/dev/arm/systemcalls.c
@@ -22,6 +22,10 @@
 #include <sys/proc_internal.h>
 #include <sys/syscall.h>
 #include <sys/systm.h>
+
+#if defined(KEBAB)
+int kebab_syscount = 0;
+#endif
 #include <sys/user.h>
 #include <sys/errno.h>
 #include <sys/kdebug.h>
@@ -92,6 +96,7 @@
 	thread_t thread_act,
 	struct proc * proc)
 {
+
 	const struct sysent  *callp;
 	int             error;
 	unsigned short  code, syscode;
@@ -102,6 +107,13 @@
 
 	code = arm_get_syscall_number(state);
 
+#if defined(KEBAB)
+	if (kebab_syscount++ < 12) {
+		printf("KEBAB: syscall %d arg0 0x%llx\n", (int)code,
+		    (unsigned long long)saved_state64(state)->x[0]);
+	}
+#endif
+
 #define unix_syscall_kprintf(x...)      /* kprintf("unix_syscall: " x) */
 
 	if (kdebug_enable && !code_is_kdebug_trace(code)) {
--- a/bsd/vfs/vfs_syscalls.c
+++ b/bsd/vfs/vfs_syscalls.c
@@ -4526,6 +4526,10 @@
 open1(vfs_context_t ctx, struct nameidata *ndp, int uflags,
     struct vnode_attr *vap, fp_initfn_t fp_init, void *initarg, int32_t *retval, int authfd)
 {
+#if defined(KEBAB)
+	printf("KEBAB: open1 enter\n");
+#endif
+
 	proc_t p = vfs_context_proc(ctx);
 	uthread_t uu = get_bsdthread_info(vfs_context_thread(ctx));
 	struct fileproc *fp;
@@ -4554,6 +4558,9 @@
 	AUDIT_ARG(fflags, oflags);
 	AUDIT_ARG(mode, vap->va_mode);
 
+#if defined(KEBAB)
+	printf("KEBAB: open1 falloc next\n");
+#endif
 	if ((error = falloc_withinit(p, &fp, &indx, ctx, fp_init, initarg)) != 0) {
 		return error;
 	}
@@ -4583,6 +4590,9 @@
 		}
 	}
 
+#if defined(KEBAB)
+	printf("KEBAB: open1 vn_open_auth next\n");
+#endif
 	if ((error = vn_open_auth(ndp, &flags, vap, authvp))) {
 		if (authvp != NULLVP) {
 			vnode_put(authvp);
--- a/bsd/vfs/vfs_vnops.c
+++ b/bsd/vfs/vfs_vnops.c
@@ -374,6 +374,10 @@
 int
 vn_open_auth(struct nameidata *ndp, int *fmodep, struct vnode_attr *vap, vnode_t authvp)
 {
+#if defined(KEBAB)
+	printf("KEBAB: vn_open_auth enter\n");
+#endif
+
 	struct vnode *vp;
 	struct vnode *dvp;
 	vfs_context_t ctx = ndp->ni_cnd.cn_context;
@@ -563,9 +567,15 @@
 
 		/* Do a lookup, possibly going directly to filesystem for compound operation */
 		do {
+#if defined(KEBAB)
+			printf("KEBAB: vn_open_auth lookup next\n");
+#endif
 			if ((error = namei(ndp))) {
 				goto out;
 			}
+#if defined(KEBAB)
+			printf("KEBAB: vn_open_auth lookup done (vp %p)\n", (void *)ndp->ni_vp);
+#endif
 			vp = ndp->ni_vp;
 			dvp = ndp->ni_dvp;
 
@@ -654,6 +664,9 @@
 			}
 		}
 
+#if defined(KEBAB)
+		printf("KEBAB: vn_open_auth VNOP_OPEN next\n");
+#endif
 		error = VNOP_OPEN(vp, fmode, ctx);
 		if (error) {
 			/*
--- a/bsd/vfs/vfs_lookup.c
+++ b/bsd/vfs/vfs_lookup.c
@@ -1089,6 +1089,7 @@
 int
 lookup(struct nameidata *ndp)
 {
+
 	char    *cp;            /* pointer into pathname argument */
 	vnode_t         tdp;            /* saved dp */
 	vnode_t         dp;             /* the directory we are searching */
@@ -1119,6 +1120,20 @@
 	dp = ndp->ni_startdir;
 	ndp->ni_startdir = NULLVP;
 
+#if defined(KEBAB)
+	{
+		char kname[32] = "?";
+		size_t klen;
+		if (cnp->cn_nameptr != NULL) {
+			for (klen = 0; klen < sizeof(kname) - 1 && klen < cnp->cn_namelen; klen++) {
+				kname[klen] = cnp->cn_nameptr[klen];
+			}
+			kname[klen] = 0;
+		}
+		printf("KEBAB: lookup \"%s\" in vp %p\n", kname, (void *)dp);
+	}
+#endif
+
 	if ((cnp->cn_flags & CN_NBMOUNTLOOK) != 0) {
 		vbusyflags = LK_NOWAIT;
 	}
